src/endpoint.rs

```diff
@@ -149,11 +241,22 @@ impl Endpoint {
             },
             Err(e) => info!("Echo service timed out: {:?}", e),
         }
-        if let Some(socket_addr) = addr {
+
+        addr.map_or(Err(Error::NoEchoServiceResponse), |socket_addr| {
+            self.public_addr = Some(socket_addr);
             Ok(socket_addr)
+        })
+    }
+
+    /// Disconnects from a peer
+    pub fn disconnect_from(&mut self, peer_addr: &SocketAddr) -> Result<()> {
+        if let Some((conn, remover)) = self.connection_pool.remove(peer_addr) {
+            conn.close(0u8.into(), b"User Disconnected");
+            remover.remove();
         } else {
-            Err(Error::NoEchoServiceResponse)
+            log::info!("We are not connected to this peer. Doing nothing");
```

Comment:
Perhaps the message should be more informative about what it was attempting to do and that it's not doing it since there is not connection, otherwise when reading the logs you cannot understand what is it that "it's not doing".

Reply:
Of course :) I was saving log clean-up for the end :smile: